### PR TITLE
Added pulpcore-manager command for removing signing services

### LINKED ISSUE #2967: Facilitate removal of signing services in functional tests
Allowing testers to call a function for removing a signing service would be useful. Currently, we remove the signing service in the following manner: https://github.com/pulp/pulpcore/blob/6287ef36d2fc20ca316e1e889296fc55072ff477/pulpcore/tests/functional/gpg_ascii_armor_signing_service.py#L158-L168

Having a fixture `remove_signing_service(service_name, service_type)` accessible globally would suffice.

### PR DESCRIPTION
**Pull request by MichalPysik**
The pulpcore-manager now supports command 'remove-signing-service' which allows for removing signing services specified by name and class.

closes #2967

### CODE REVIEW

File: pulpcore/app/management/commands/remove-signing-service.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+        if ":" not in options["class"]:
+            raise CommandError(_("The signing service class was not provided in a proper format."))
+        app_label, service_class = options["class"].split(":")
+
+        try:
+            SigningService = apps.get_model(app_label, service_class)
+        except LookupError as e:
+            raise CommandError(str(e))
+
+        available_signing_services = {ss.name for ss in SigningService.objects.all()}
```
Comment (mdellweg) [inline pulpcore/app/management/commands/remove-signing-service.py:41]:
```python
service = SigningService.objects.get(name=name)
service.delete()
```
Should have been sufficient. You'd catch both `IntegrityError` and `DoesNotExist`.

Comment (MichalPysik) [reply pulpcore/app/management/commands/remove-signing-service.py:41]: That sounds good, I will change that for the deletion code. Although I will keep the lines you selected (38-41), since their purpose is mainly to show the user the list of all available services when he selects a non-existent one.

File: pulpcore/tests/functional/gpg_ascii_armor_signing_service.py [hunk 2]
```diff
@@ -157,9 +157,10 @@ def _ascii_armored_detached_signing_service_name(
 
     cmd = (
         "pulpcore-manager",
-        "shell",
-        "-c",
-        f'from pulpcore.app.models import AsciiArmoredDetachedSigningService;print(AsciiArmoredDetachedSigningService.objects.get(name="{service_name}").delete())',  # noqa: E501
+        "remove-signing-service",
+        service_name,
+        "--class",
+        "core:AsciiArmoredDetachedSigningService",
```
Comment (mdellweg) [inline pulpcore/tests/functional/gpg_ascii_armor_signing_service.py:163]: :+1: Thank you for not forgetting this!

File: pulpcore/app/management/commands/remove-signing-service.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+                    "For more information, please refer to the detailed error message below:"
+                ).format(name)
+            )
+            raise e
+        except ObjectDoesNotExist as e:
+            self.stdout.write(
+                _(
+                    "Model: '{}' does not exist, which probably means that it was deleted by "
+                    "a different process while this action was in progress. "
+                    "For more information, please refer to the detailed error message below:"
```
Comment (mdellweg) [inline pulpcore/app/management/commands/remove-signing-service.py:69]: I would rephrase this to `raise CommandError("Signing service {name} of class {} does not exist.").
Reason: Let's not start to speculate for the user. The name could have been a typo and never existed anyway. Also the exception tells us "The thing de not exist". That is all the information the user ever wants to hear. There is no point in rerasining it.

File: pulpcore/app/management/commands/remove-signing-service.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+            SigningService.objects.get(name=name).delete()
+        except IntegrityError as e:
+            self.stdout.write(
+                _(
+                    "Model: '{}' could not be removed, because it is being referenced by a "
+                    "different model through a relationship that preserves the database integrity. "
+                    "For more information, please refer to the detailed error message below:"
+                ).format(name)
+            )
+            raise e
```
Comment (mdellweg) [inline pulpcore/app/management/commands/remove-signing-service.py:63]: Same here. Don't generate too much information: `raise CommandError("SingingService could not be removed because it is still in use.")`

File: pulpcore/app/management/commands/remove-signing-service.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+            raise CommandError(str(e))
+
+        available_signing_services = {ss.name for ss in SigningService.objects.all()}
+        if name not in available_signing_services:
+            raise CommandError(
+                (
+                    "Signing service name is either incorrectly specified or does not exist. "
+                    "Please specify one of the following signing service names: {}."
+                ).format(list(available_signing_services))
+            )
```
Comment (mdellweg) [inline pulpcore/app/management/commands/remove-signing-service.py:50]: Let's not do this. The user didn't ask for it. There's an api to list signing services.

File: pulpcore/app/management/commands/remove-signing-service.py (hunk 3 above)
Comment (MichalPysik) [reply pulpcore/app/management/commands/remove-signing-service.py:69]: The problem with this is that it could not have been a typo, since if the service didn't exist, it wouldn't get found in the list of all available services above. But I can still shorted the message tho

File: pulpcore/app/management/commands/remove-signing-service.py [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+        except LookupError as e:
+            raise CommandError(str(e))
+
+        try:
+            SigningService.objects.get(name=name).delete()
+        except IntegrityError as e:
+            self.stdout.write(
+                _(
+                    "Signing service '{}' could not be removed because it's still in use. "
+                    "For more information, see the detailed error message below:"
```
Comment (mdellweg) [inline pulpcore/app/management/commands/remove-signing-service.py:49]: Can you give an example of the "more information" please?

File: pulpcore/app/management/commands/remove-signing-service.py [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+                ).format(name)
+            )
+            raise e
+        except ObjectDoesNotExist as e:
+            self.stdout.write(
+                _(
+                    "Signing service '{}' of class '{}' does not exists. "
+                    "For more information, see the detailed error message below:"
+                ).format(name)
```
Comment (mdellweg) [inline pulpcore/app/management/commands/remove-signing-service.py:58]: I really do not expect more information here. The thing does not exist. That is it.

File: pulpcore/app/management/commands/remove-signing-service.py [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+
+        try:
+            SigningService.objects.get(name=name).delete()
+        except IntegrityError as e:
+            self.stdout.write(
+                _("Signing service '{}' could not be removed because it's still in use.").format(
+                    name
+                )
+            )
+            raise e
```
Comment (mdellweg) [inline pulpcore/app/management/commands/remove-signing-service.py:51]:
```suggestion
            raise CommandError(
                _("Signing service '{}' could not be removed because it's still in use.").format(
                    name
                )
            )
```

File: pulpcore/app/management/commands/remove-signing-service.py [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+        except ObjectDoesNotExist as e:
+            self.stdout.write(_("Signing service '{}' of class '{}' does not exists.").format(name))
+            raise e
```
Comment (mdellweg) [inline pulpcore/app/management/commands/remove-signing-service.py:54]:
```suggestion
            raise CommandError(_("Signing service '{}' of class '{}' does not exists.").format(name))
```

File: pulpcore/app/management/commands/remove-signing-service.py [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+        name = options["name"]
+        if ":" not in options["class"]:
+            raise CommandError(_("The signing service class was not provided in a proper format."))
+        app_label, service_class = options["class"].split(":")
+
+        try:
+            SigningService = apps.get_model(app_label, service_class)
+        except LookupError as e:
+            raise CommandError(str(e))
+
```
Comment (mdellweg) [inline pulpcore/app/management/commands/remove-signing-service.py:42]: Can we add a check here that the class is a subclass of the signing service base class?

**Comment by MichalPysik** [comment]
Should I maybe rewrite the stdout writes to raises of CommandError instead? @mdellweg

File: pulpcore/app/management/commands/remove-signing-service.py [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+        name = options["name"]
+        if ":" not in options["class"]:
+            raise CommandError(_("The signing service class was not provided in a proper format."))
+        app_label, service_class = options["class"].split(":")
+
+        try:
+            SigningService = apps.get_model(app_label, service_class)
+        except LookupError as e:
+            raise CommandError(str(e))
+        if SigningService.__base__ is not BaseSigningService:
```
Comment (mdellweg) [inline pulpcore/app/management/commands/remove-signing-service.py:44]: there's a builtin `issubclass`.
Also can you add the same check to the other management command?

Comment (MichalPysik) [reply pulpcore/app/management/commands/remove-signing-service.py:44]: I tried that, but it didn't work, maybe I did it incorrectly so I will try it again. What other command do you mean, ``add-signing-service`` ?

**Review by mdellweg — APPROVED** [review]

File: CHANGES/2967.feature [hunk 12]
```diff
@@ -0,0 +1 @@
+Added pulpcore-manager command called remove-signing-service for removing specified signing services.
```
Comment (mdellweg) [inline CHANGES/2967.feature:1]:
```suggestion
Added pulpcore-manager command called ``remove-signing-service`` for removing specified signing services.
```

File: pulpcore/tests/functional/gpg_ascii_armor_signing_service.py [hunk 13]
```diff
@@ -157,9 +157,10 @@ def _ascii_armored_detached_signing_service_name(
 
     cmd = (
         "pulpcore-manager",
-        "shell",
```
Comment (bmbouter) [inline pulpcore/tests/functional/gpg_ascii_armor_signing_service.py:160]: :+1:

File: pulpcore/app/management/commands/remove-signing-service.py [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+        if not issubclass(SigningService, BaseSigningService):
+            raise CommandError(
+                _("Class '{}' is not a subclass of the base 'core:SigningService' class.").format(
+                    options["class"]
+                )
+            )
+
+        try:
+            SigningService.objects.get(name=name).delete()
+            self.stdout.write(_("Signing service '{}' has been successfully removed.").format(name))
```
Comment (bmbouter) [inline pulpcore/app/management/commands/remove-signing-service.py:53]: This could move out of the try block entirely or to an `else` clause.

**Review by bmbouter — CHANGES_REQUESTED** [review]
@MichalPysik right now we have fixtures that use these commands, but can you add a test in pulpcore that creates and deletes a `core:AsciiArmoredDetachedSigningService` signing service. I just did an audit in pulpcore and pulp_file and I don't see any test that actually uses these fixtures in either repo. Since the signing service and fixture are in pulpcore, I think it can just live in pulpcore.

What do you think? If this doesn't make sense, please ask. Thanks!

File: pulpcore/tests/functional/api/test_signing_service.py [hunk 15]
```diff
@@ -0,0 +1,8 @@
+import pytest
+
+
+@pytest.mark.parallel
+def test_crud_signing_service(ascii_armored_detached_signing_service):
+    service = ascii_armored_detached_signing_service
+    assert service
+    assert service.pulp_href.isascii()
```
Comment (mdellweg) [inline pulpcore/tests/functional/api/test_signing_service.py:8]: I thought that you'd check that the pulp_href contains the string "ascii". And maybe it doesn't...

File: pulpcore/tests/functional/api/test_signing_service.py [hunk 16]
```diff
@@ -0,0 +1,7 @@
+import pytest
+
+
+@pytest.mark.parallel
+def test_crud_signing_service(ascii_armored_detached_signing_service):
```
Comment (bmbouter) [inline pulpcore/tests/functional/api/test_signing_service.py:5]: :heart_eyes:

**Review by bmbouter — APPROVED** [review]
Thank you!!

**Comment by bmbouter** [comment]
I see two approvals so I'm going to merge.

File: pulpcore/tests/functional/api/test_signing_service.py (hunk 15 above)
Comment (MichalPysik) [reply pulpcore/tests/functional/api/test_signing_service.py]: The pulp_href doesn't contain the word 'ascii', I checked the format of the href.

**Comment by tjmullicani** [comment]
Any idea when this feature will be integrated into the pulpcore release schedule? I didn't see it in 3.21.3.

**Comment by dralley** [comment]
It'll be in the next released, probably 1-2 weeks from now.
